beacon/types/exec_header.go

```diff
@@ -48,33 +52,40 @@ func ExecutionHeaderFromJSON(forkName string, data []byte) (*ExecutionHeader, er
 	default:
 		return nil, fmt.Errorf("unsupported fork: %s", forkName)
 	}
-	if err := json.Unmarshal(data, obj); err != nil {
-		return nil, err
+
+	// Use a streaming decoder for efficiency, especially with large JSON payloads
+	decoder := json.NewDecoder(bytes.NewReader(data))
+	if err := decoder.Decode(obj); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal execution header for fork %s: %w", forkName, err)
 	}
 	return &ExecutionHeader{obj: obj}, nil
 }
 
+// NewExecutionHeader initializes a new ExecutionHeader object with the given headerObject.
+// It ensures the object type is one of the supported ExecutionPayloadHeaders (capella, deneb).
 func NewExecutionHeader(obj headerObject) *ExecutionHeader {
 	switch obj.(type) {
-	case *capella.ExecutionPayloadHeader:
-	case *deneb.ExecutionPayloadHeader:
+	case *capella.ExecutionPayloadHeader, *deneb.ExecutionPayloadHeader:
+		// Supported types
 	default:
 		panic(fmt.Errorf("unsupported ExecutionPayloadHeader type %T", obj))
 	}
 	return &ExecutionHeader{obj: obj}
 }
 
+// PayloadRoot returns the Merkle root of the execution payload header.
 func (eh *ExecutionHeader) PayloadRoot() merkle.Value {
 	return merkle.Value(eh.obj.HashTreeRoot(tree.GetHashFn()))
 }
 
+// BlockHash extracts the block hash from the underlying execution payload header.
+// It checks the type of the object and returns the correct block hash based on the type.
 func (eh *ExecutionHeader) BlockHash() common.Hash {
-	switch obj := eh.obj.(type) {
-	case *capella.ExecutionPayloadHeader:
+	if obj, ok := eh.obj.(*capella.ExecutionPayloadHeader); ok {
```

Comment:
meh, the earlier code was better, IMO, written for future extensibility

Reply:
Thanks for the feedback! I can see your point—using the switch statement does make it more extensible for future cases. The type assertion approach was primarily intended for readability and simplicity in the current context. I'll switch back to the previous approach to maintain future extensibility, especially if additional types are introduced later.

Reply:
Thanks for the reply! I can see your point—using the switch statement does make it more extensible for future cases. The type assertion approach was primarily intended for readability and simplicity in the current context. I'll switch back to the previous approach to maintain future extensibility, especially if additional types are introduced later.